### _Imports_

In [1]:
import pandas as pd
import numpy as np

pd.set_option('display.max_columns', None)

In [2]:
df = pd.read_csv('../data/Sample - Superstore.csv', encoding='ISO-8859-1')
df.head(10)

Row ID        Order ID  Order Date   Ship Date       Ship Mode Customer ID  \
0       1  CA-2016-152156   11/8/2016  11/11/2016    Second Class    CG-12520   
1       2  CA-2016-152156   11/8/2016  11/11/2016    Second Class    CG-12520   
2       3  CA-2016-138688   6/12/2016   6/16/2016    Second Class    DV-13045   
3       4  US-2015-108966  10/11/2015  10/18/2015  Standard Class    SO-20335   
4       5  US-2015-108966  10/11/2015  10/18/2015  Standard Class    SO-20335   
5       6  CA-2014-115812    6/9/2014   6/14/2014  Standard Class    BH-11710   
6       7  CA-2014-115812    6/9/2014   6/14/2014  Standard Class    BH-11710   
7       8  CA-2014-115812    6/9/2014   6/14/2014  Standard Class    BH-11710   
8       9  CA-2014-115812    6/9/2014   6/14/2014  Standard Class    BH-11710   
9      10  CA-2014-115812    6/9/2014   6/14/2014  Standard Class    BH-11710   

     Customer Name    Segment        Country             City       State  \
0      Claire Gute   Consumer  United States        Henderson    Kentucky   
1      Claire Gute   Consumer  United States        Henderson    Kentucky   
2  Darrin Van Huff  Corporate  United States      Los Angeles  California   
3   Sean O'Donnell   Consumer  United States  Fort Lauderdale     Florida   
4   Sean O'Donnell   Consumer  United States  Fort Lauderdale     Florida   
5  Brosina Hoffman   Consumer  United States      Los Angeles  California   
6  Brosina Hoffman   Consumer  United States      Los Angeles  California   
7  Brosina Hoffman   Consumer  United States      Los Angeles  California   
8  Brosina Hoffman   Consumer  United States      Los Angeles  California   
9  Brosina Hoffman   Consumer  United States      Los Angeles  California   

   Postal Code Region       Product ID         Category Sub-Category  \
0        42420  South  FUR-BO-10001798        Furniture    Bookcases   
1        42420  South  FUR-CH-10000454        Furniture       Chairs   
2        90036   West  OFF-LA-10000240  Office Supplies       Labels   
3        33311  South  FUR-TA-10000577        Furniture       Tables   
4        33311  South  OFF-ST-10000760  Office Supplies      Storage   
5        90032   West  FUR-FU-10001487        Furniture  Furnishings   
6        90032   West  OFF-AR-10002833  Office Supplies          Art   
7        90032   West  TEC-PH-10002275       Technology       Phones   
8        90032   West  OFF-BI-10003910  Office Supplies      Binders   
9        90032   West  OFF-AP-10002892  Office Supplies   Appliances   

                                        Product Name     Sales  Quantity  \
0                  Bush Somerset Collection Bookcase  261.9600         2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400         3   
2  Self-Adhesive Address Labels for Typewriters b...   14.6200         2   
3      Bretford CR4500 Series Slim Rectangular Table  957.5775         5   
4                     Eldon Fold 'N Roll Cart System   22.3680         2   
5  Eldon Expressions Wood and Plastic Desk Access...   48.8600         7   
6                                         Newell 322    7.2800         4   
7                     Mitel 5320 IP Phone VoIP phone  907.1520         6   
8  DXL Angle-View Binders with Locking Rings by S...   18.5040         3   
9                   Belkin F5C206VTEL 6 Outlet Surge  114.9000         5   

   Discount    Profit  
0      0.00   41.9136  
1      0.00  219.5820  
2      0.00    6.8714  
3      0.45 -383.0310  
4      0.20    2.5164  
5      0.00   14.1694  
6      0.00    1.9656  
7      0.20   90.7152  
8      0.20    5.7825  
9      0.00   34.4700

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row ID         9994 non-null   int64  
 1   Order ID       9994 non-null   object 
 2   Order Date     9994 non-null   object 
 3   Ship Date      9994 non-null   object 
 4   Ship Mode      9994 non-null   object 
 5   Customer ID    9994 non-null   object 
 6   Customer Name  9994 non-null   object 
 7   Segment        9994 non-null   object 
 8   Country        9994 non-null   object 
 9   City           9994 non-null   object 
 10  State          9994 non-null   object 
 11  Postal Code    9994 non-null   int64  
 12  Region         9994 non-null   object 
 13  Product ID     9994 non-null   object 
 14  Category       9994 non-null   object 
 15  Sub-Category   9994 non-null   object 
 16  Product Name   9994 non-null   object 
 17  Sales          9994 non-null   float64
 18  Quantity

In [4]:
# Good news, there are no nulls

The aim of this notebook is to check whether we can build a normalized data model with this dataset.
Hereby one example of data model for similar dataset:

![Screenshot](support_doc/ExampleDataModel.png)

Lets define the tables with the corresponding features:  

**orders**:  
*order_id  
geography_id  
customer_id  
order_date  
ship_date  
ship_mode  

**geography**  
*geography_id  ⇒ not in the dataset  
postal_code  
city  
region  
state  
country  

**customers**:  
*customer_id  
name  
segment

**products**  
*product_id  
category  
subcategory  
name  

**order_items**  
*order_id  
*item_id ⇒ 'Row ID'    
product_id  
quantity  
sales  
discount  
profit  


Now lets see if the relationships between variables are as we described them above.

In [5]:
# orders

cols_lst = ['Customer ID', 'Order Date', 'Ship Date', 'Ship Mode']
primary_key = 'Order ID'
tab_name = 'orders'

all_unique = True
for col in cols_lst:
    counts = df.groupby(primary_key)[col].nunique()
    max_count = counts.max()
    if max_count !=1:
        print(f'{col} is not unique in {tab_name} table')
        all_unique = False
if all_unique:
    print(f'All columns are unique in {tab_name} table')

All columns are unique in orders table


In [6]:
# geography

cols_lst = ['City', 'Region', 'State', 'Country']
primary_key = 'Postal Code'
tab_name = 'geography'

all_unique = True
for col in cols_lst:
    counts = df.groupby(primary_key)[col].nunique()
    max_count = counts.max()
    if max_count !=1:
        print(f'{col} is not unique in {tab_name} table')
        all_unique = False
if all_unique:
    print(f'All columns are unique in {tab_name} table')

City is not unique in geography table


In [7]:
# Lets check how many rows are affected

counts = df.groupby(primary_key)['City'].nunique()
counts.sort_values(ascending=False, inplace=True)
counts

Postal Code
92024    2
1040     1
76021    1
76063    1
76106    1
        ..
40324    1
40475    1
41042    1
42071    1
99301    1
Name: City, Length: 631, dtype: int64

In [8]:
check = df[df['Postal Code'] == 92024]
check

Row ID        Order ID  Order Date   Ship Date       Ship Mode  \
481      482  CA-2014-134677   10/6/2014  10/10/2014  Standard Class   
594      595  CA-2014-131450    8/8/2014   8/15/2014  Standard Class   
595      596  CA-2014-131450    8/8/2014   8/15/2014  Standard Class   
596      597  CA-2014-131450    8/8/2014   8/15/2014  Standard Class   
597      598  CA-2014-131450    8/8/2014   8/15/2014  Standard Class   
1113    1114  CA-2017-140585  12/18/2017  12/23/2017    Second Class   
1114    1115  CA-2017-140585  12/18/2017  12/23/2017    Second Class   
1115    1116  CA-2017-140585  12/18/2017  12/23/2017    Second Class   
1543    1544  CA-2014-107181    2/4/2014    2/8/2014  Standard Class   
1544    1545  CA-2014-107181    2/4/2014    2/8/2014  Standard Class   
1551    1552  CA-2016-159345   6/17/2016   6/22/2016  Standard Class   
1924    1925  CA-2017-126865   12/5/2017   12/7/2017    Second Class   
2259    2260  CA-2015-117611   11/8/2015  11/10/2015    Second Class   
2260    2261  CA-2015-117611   11/8/2015  11/10/2015    Second Class   
2606    2607  CA-2016-154053    3/4/2016   3/10/2016  Standard Class   
2674    2675  CA-2017-136875   12/3/2017   12/3/2017        Same Day   
2819    2820  CA-2016-160500    5/5/2016    5/8/2016     First Class   
2820    2821  CA-2016-160500    5/5/2016    5/8/2016     First Class   
2946    2947  CA-2017-169859  12/14/2017  12/18/2017  Standard Class   
2947    2948  CA-2017-169859  12/14/2017  12/18/2017  Standard Class   
2948    2949  CA-2017-169859  12/14/2017  12/18/2017  Standard Class   
3292    3293  CA-2014-106264  12/26/2014  12/30/2014  Standard Class   
3293    3294  CA-2014-106264  12/26/2014  12/30/2014  Standard Class   
3528    3529  CA-2016-130680   6/24/2016   6/29/2016  Standard Class   
4396    4397  CA-2016-148740  11/15/2016  11/19/2016  Standard Class   
5638    5639  CA-2014-154893  12/21/2014  12/27/2014  Standard Class   
5639    5640  CA-2014-154893  12/21/2014  12/27/2014  Standard Class   
6298    6299  CA-2014-169852  10/21/2014  10/26/2014  Standard Class   
6672    6673  CA-2015-153794   9/11/2015   9/16/2015  Standard Class   
6698    6699  CA-2017-145429   7/21/2017   7/25/2017  Standard Class   
6699    6700  CA-2017-145429   7/21/2017   7/25/2017  Standard Class   
6700    6701  CA-2017-145429   7/21/2017   7/25/2017  Standard Class   
7189    7190  CA-2016-164399  11/12/2016  11/15/2016     First Class   
7190    7191  CA-2016-164399  11/12/2016  11/15/2016     First Class   
7918    7919  CA-2015-152513    7/4/2015    7/8/2015    Second Class   
8684    8685  CA-2016-155446  12/16/2016  12/23/2016  Standard Class   
8955    8956  CA-2014-126480   9/20/2014   9/24/2014  Standard Class   
9060    9061  CA-2015-136420  11/20/2015  11/21/2015     First Class   
9061    9062  CA-2015-136420  11/20/2015  11/21/2015     First Class   

     Customer ID      Customer Name      Segment        Country       City  \
481     XP-21865       Xylona Preis     Consumer  United States  San Diego   
594     LR-16915       Lena Radford     Consumer  United States  San Diego   
595     LR-16915       Lena Radford     Consumer  United States  San Diego   
596     LR-16915       Lena Radford     Consumer  United States  San Diego   
597     LR-16915       Lena Radford     Consumer  United States  San Diego   
1113    RA-19915  Russell Applegate     Consumer  United States  Encinitas   
1114    RA-19915  Russell Applegate     Consumer  United States  Encinitas   
1115    RA-19915  Russell Applegate     Consumer  United States  Encinitas   
1543    DB-13270  Deborah Brumfield  Home Office  United States  San Diego   
1544    DB-13270  Deborah Brumfield  Home Office  United States  San Diego   
1551    IG-15085        Ivan Gibson     Consumer  United States  San Diego   
1924    NP-18325       Naresj Patel     Consumer  United States  San Diego   
2259    MZ-17335      Maria Zettner  Home Office  United States  San Diego   
2260    MZ-17335      Maria Zettn

In [9]:
check['City'].value_counts()

City
San Diego    34
Encinitas     5
Name: count, dtype: int64

It seems like an error. We can replace Encinitas by San Diego.

In [52]:
# Our assumption is that geography is related to order. Let's double-check that there is no 1:1 relationship with customer

counts = df.groupby('Customer ID')['Postal Code'].nunique()

counts.sort_values()

Customer ID
RE-19405     1
TC-21145     1
AR-10570     1
SM-20905     1
HH-15010     1
            ..
ZC-21910    13
JE-15745    13
EA-14035    13
SH-19975    13
EP-13915    17
Name: Postal Code, Length: 793, dtype: int64

Conclusion: our assumption is right. Geography columns are related to the order placement.

In [10]:
# customers

cols_lst = ['Customer Name', 'Segment']
primary_key = 'Customer ID'
tab_name = 'customers'

all_unique = True
for col in cols_lst:
    counts = df.groupby(primary_key)[col].nunique()
    max_count = counts.max()
    if max_count !=1:
        print(f'{col} is not unique in {tab_name} table')
        all_unique = False
if all_unique:
    print(f'All columns are unique in {tab_name} table')

All columns are unique in customers table


In [11]:
# products

cols_lst = ['Category', 'Sub-Category', 'Product Name']
primary_key = 'Product ID'
tab_name = 'products'

all_unique = True
for col in cols_lst:
    counts = df.groupby(primary_key)[col].nunique()
    max_count = counts.max()
    if max_count !=1:
        print(f'{col} is not unique in {tab_name} table')
        all_unique = False
if all_unique:
    print(f'All columns are unique in {tab_name} table')

Product Name is not unique in products table


In [22]:
# Lets check how many rows are affected

counts = df.groupby(primary_key)['City'].nunique()
counts.sort_values(ascending=False, inplace=True)
counts

Product ID
FUR-FU-10004270    16
OFF-PA-10001970    14
FUR-CH-10002647    14
TEC-AC-10003832    13
OFF-BI-10004632    13
                   ..
TEC-MA-10002412     1
TEC-MA-10002694     1
TEC-MA-10002790     1
TEC-MA-10002927     1
FUR-BO-10000112     1
Name: City, Length: 1862, dtype: int64

In [13]:
check = df[df['Product ID'] == 'FUR-FU-10004270']
check

Row ID        Order ID  Order Date   Ship Date       Ship Mode  \
1473    1474  US-2015-105676   12/1/2015   12/2/2015        Same Day   
1509    1510  CA-2017-103611   9/12/2017   9/15/2017     First Class   
1801    1802  CA-2016-149461  11/13/2016  11/19/2016  Standard Class   
2645    2646  CA-2014-131002    9/7/2014   9/12/2014    Second Class   
2694    2695  CA-2014-143182  10/15/2014  10/20/2014  Standard Class   
2700    2701  CA-2014-145317   3/18/2014   3/23/2014  Standard Class   
2758    2759  CA-2016-146171   3/11/2016   3/15/2016  Standard Class   
3040    3041  US-2016-147991    5/5/2016    5/9/2016  Standard Class   
4247    4248  CA-2017-117702  11/28/2017   12/4/2017  Standard Class   
4853    4854  CA-2014-113320  12/12/2014  12/15/2014    Second Class   
5325    5326  CA-2015-119508   12/4/2015   12/9/2015  Standard Class   
5757    5758  CA-2015-119634   8/11/2015   8/16/2015  Standard Class   
7287    7288  CA-2016-149965   6/20/2016   6/25/2016  Standard Class   
7886    7887  CA-2017-118017   12/3/2017   12/6/2017    Second Class   
8030    8031  CA-2016-158806    1/7/2016   1/11/2016  Standard Class   
8676    8677  CA-2017-163265   2/16/2017   2/21/2017  Standard Class   

     Customer ID       Customer Name      Segment        Country  \
1473    NM-18520        Neoma Murray     Consumer  United States   
1509    JM-15535      Jessica Myrick     Consumer  United States   
1801    AS-10135        Adrian Shami  Home Office  United States   
2645    TB-21400    Tom Boeckenhauer     Consumer  United States   
2694    DL-12865          Dan Lawera     Consumer  United States   
2700    SM-20320         Sean Miller  Home Office  United States   
2758    JP-16135      Julie Prescott  Home Office  United States   
3040    ZD-21925  Zuschuss Donatelli     Consumer  United States   
4247    LS-16975    Lindsay Shagiari  Home Office  United States   
4853    LH-17155    Logan Haushalter     Consumer  United States   
5325    TZ-21580           Tracy Zic     Consumer  United States   
5757    BW-11065       Barry Weirich     Consumer  United States   
7287    BS-11365        Bill Shonely    Corporate  United States   
7886    LC-16870       Lena Cacioppo     Consumer  United States   
8030    NM-18520        Neoma Murray     Consumer  United States   
8676    JS-16030           Joy Smith     Consumer  United States   

               City           State  Postal Code   Region       Product ID  \
1473        Houston           Texas        77036  Central  FUR-FU-10004270   
1509    Los Angeles      California        90036     West  FUR-FU-10004270   
1801         Auburn      Washington        98002     West  FUR-FU-10004270   
2645          Tulsa        Oklahoma        74133  Central  FUR-FU-10004270   
2694        Hialeah         Florida        33012    South  FUR-FU-10004270   
2700   Jacksonville         Florida        32216    South  FUR-FU-10004270   
2758       Columbus         Georgia        31907    South  FUR-FU-10004270   
3040    Chattanooga       Tennessee        37421    South  FUR-FU-10004270   
4247      Baltimore        Maryland        21215     East  FUR-FU-10004270   
4853        Oakland      California        94601     West  FUR-FU-10004270   
5325       Lakewood      California        90712     West  FUR-FU-10004270   
5757        Raleigh  North Carolina        27604    South  FUR-FU-10004270   
7287  Oklahoma City        Oklahoma        73120  Central  FUR-FU-10004270   
7886       Thornton        Colorado        80229     West  FUR-FU-10004270   
8030       Amarillo           Texas        79109  Central  FUR-FU-10004270   
8676        Decatur        Illinois        62521  Central  FUR-FU-10004270   

       Category Sub-Category                                    Product Name  \
1473  Furniture  Furnishings   Eldon Image Series Desk Accessories, Burgundy   
1509  Furniture  Furnishings   Eldon Image Series Desk Accessories, Burgundy   
1801  Furniture  Furnishings   Eldon Image Series Desk Ac

In [14]:
check['Product Name'].value_counts()

Product Name
Eldon Image Series Desk Accessories, Burgundy     8
Executive Impressions 13" Clairmont Wall Clock    8
Name: count, dtype: int64

The prices for these products are different, therefore we can correct the id of one of the products. But first lets check whether the different products are related to different dates/years

In [15]:
# Changin the format of Order and Shipe Date to 'datetime'

check['Order Date'] = pd.to_datetime(check['Order Date'])
check['Ship Date'] = pd.to_datetime(check['Ship Date'])

C:\Users\Kristina\AppData\Local\Temp\ipykernel_13952\1994163358.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  check['Order Date'] = pd.to_datetime(check['Order Date'])
C:\Users\Kristina\AppData\Local\Temp\ipykernel_13952\1994163358.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  check['Ship Date'] = pd.to_datetime(check['Ship Date'])


In [16]:
# Sorting the dataframe by date

check.sort_values(by = 'Order Date')

Row ID        Order ID Order Date  Ship Date       Ship Mode  \
2700    2701  CA-2014-145317 2014-03-18 2014-03-23  Standard Class   
2645    2646  CA-2014-131002 2014-09-07 2014-09-12    Second Class   
2694    2695  CA-2014-143182 2014-10-15 2014-10-20  Standard Class   
4853    4854  CA-2014-113320 2014-12-12 2014-12-15    Second Class   
5757    5758  CA-2015-119634 2015-08-11 2015-08-16  Standard Class   
1473    1474  US-2015-105676 2015-12-01 2015-12-02        Same Day   
5325    5326  CA-2015-119508 2015-12-04 2015-12-09  Standard Class   
8030    8031  CA-2016-158806 2016-01-07 2016-01-11  Standard Class   
2758    2759  CA-2016-146171 2016-03-11 2016-03-15  Standard Class   
3040    3041  US-2016-147991 2016-05-05 2016-05-09  Standard Class   
7287    7288  CA-2016-149965 2016-06-20 2016-06-25  Standard Class   
1801    1802  CA-2016-149461 2016-11-13 2016-11-19  Standard Class   
8676    8677  CA-2017-163265 2017-02-16 2017-02-21  Standard Class   
1509    1510  CA-2017-103611 2017-09-12 2017-09-15     First Class   
4247    4248  CA-2017-117702 2017-11-28 2017-12-04  Standard Class   
7886    7887  CA-2017-118017 2017-12-03 2017-12-06    Second Class   

     Customer ID       Customer Name      Segment        Country  \
2700    SM-20320         Sean Miller  Home Office  United States   
2645    TB-21400    Tom Boeckenhauer     Consumer  United States   
2694    DL-12865          Dan Lawera     Consumer  United States   
4853    LH-17155    Logan Haushalter     Consumer  United States   
5757    BW-11065       Barry Weirich     Consumer  United States   
1473    NM-18520        Neoma Murray     Consumer  United States   
5325    TZ-21580           Tracy Zic     Consumer  United States   
8030    NM-18520        Neoma Murray     Consumer  United States   
2758    JP-16135      Julie Prescott  Home Office  United States   
3040    ZD-21925  Zuschuss Donatelli     Consumer  United States   
7287    BS-11365        Bill Shonely    Corporate  United States   
1801    AS-10135        Adrian Shami  Home Office  United States   
8676    JS-16030           Joy Smith     Consumer  United States   
1509    JM-15535      Jessica Myrick     Consumer  United States   
4247    LS-16975    Lindsay Shagiari  Home Office  United States   
7886    LC-16870       Lena Cacioppo     Consumer  United States   

               City           State  Postal Code   Region       Product ID  \
2700   Jacksonville         Florida        32216    South  FUR-FU-10004270   
2645          Tulsa        Oklahoma        74133  Central  FUR-FU-10004270   
2694        Hialeah         Florida        33012    South  FUR-FU-10004270   
4853        Oakland      California        94601     West  FUR-FU-10004270   
5757        Raleigh  North Carolina        27604    South  FUR-FU-10004270   
1473        Houston           Texas        77036  Central  FUR-FU-10004270   
5325       Lakewood      California        90712     West  FUR-FU-10004270   
8030       Amarillo           Texas        79109  Central  FUR-FU-10004270   
2758       Columbus         Georgia        31907    South  FUR-FU-10004270   
3040    Chattanooga       Tennessee        37421    South  FUR-FU-10004270   
7287  Oklahoma City        Oklahoma        73120  Central  FUR-FU-10004270   
1801         Auburn      Washington        98002     West  FUR-FU-10004270   
8676        Decatur        Illinois        62521  Central  FUR-FU-10004270   
1509    Los Angeles      California        90036     West  FUR-FU-10004270   
4247      Baltimore        Maryland        21215     East  FUR-FU-10004270   
7886       Thornton        Colorado        80229     West  FUR-FU-10004270   

       Category Sub-Category                                    Product Name  \
2700  Furniture  Furnishings  Executive Impressions 13" Clairmont Wall Clock   
2645  Furniture  Furnishings  Executive Impressions 13" Clairmont Wall Clock   
2694  Furniture  Furnishings  Executive Impressions 13" Clairmont Wall Clock   
4853  Furni

The date doesn't seem to be an issue. So, this is an error in the dataset and we will need to change product identifier. To define the right strategy, lets see how many products are affected.

In [41]:
counts.head(1765)

Product ID
FUR-FU-10004270    16
OFF-PA-10001970    14
FUR-CH-10002647    14
TEC-AC-10003832    13
OFF-BI-10004632    13
                   ..
FUR-CH-10002758     2
OFF-FA-10001754     2
TEC-PH-10004875     1
TEC-PH-10004830     1
FUR-BO-10001567     1
Name: City, Length: 1765, dtype: int64

1762 product codes out of 1862 are not unique. Therefore, we will need to set a new product identifier as a primary key for products table.

In [43]:
# order items

cols_lst = ['Order ID', 'Product ID', 'Quantity', 'Sales', 'Discount', 'Profit']
primary_key = 'Row ID'
tab_name = 'order_items'

Since the order items correspond to each row, each order item is unique.

In [44]:
df['Row ID'].unique()

array([   1,    2,    3, ..., 9992, 9993, 9994], dtype=int64)

**CONCLUSION**: to build a normalized data model we need to solve two issues:
1. For postal code 92024, replace City by San Diego
2. Create new Product ID, unique for each product name

Additionally, we need to create Geography ID in geography table, based on postal code.